### ISSUE TITLE
[css-contain] Container queries within display:none are difficult to implement

### ISSUE THREAD

**Issue by Loirooriol** [w3c/csswg-drafts#8197]
```html
<style>
@container (width >= 0) {
  #target { --x: BAD; }
}
</style>
<div id="outer-container" style="container-type: inline-size; --x: GOOD">
  <div id="none" style="display: none">
    <div id="inner-container" style="container-type: inline-size">
      <div id="target"></div>
    </div>
  </div>
</div>
<script>
document.body.append(getComputedStyle(target).getPropertyValue('--x'));
</script>
```

The query container for `#target` is `#inner-container`, since it has the right `container-type`. However, it generates no box since it's inside a `display: none` subtree, so `width >= 0` should be unknown and `--x: BAD` should not apply.

This is tricky to implement, though, because browsers typically avoid computing the style of elements inside `display: none` subtrees. So when collecting the rules for `#target`, we may not know the `container-type` for `#inner-container`, and thus not know whether to apply `--x: BAD`.

Currently,
 - WebKit implements it correctly.
 - Blink implements it partially: the testcase above is good, but adding this breaks it:
    ```js
    getComputedStyle(document.getElementById("inner-container")).getPropertyValue('--x');
    ```
 - Firefox skips `#inner-container` and thinks that the query container is `#outer-container`.

In https://phabricator.services.mozilla.com/D164113 @emilio says

> Our behavior doesn't seem too terrible to me. Maybe elements inside `display: none` subtrees should be considered to not have any container? Or maybe we should always match the query as "unknown", as that's really what's going on? I don't know.

**Comment by emilio** [comment]
For the record, WebKit might implement this particular case correctly, but has similar issues, see [bug 186882](https://bugs.webkit.org/show_bug.cgi?id=186882).

**Comment by emilio** [comment]
Oh, though that was fixed, at some point, so nvm then. WebKit still fails [this test](http://wpt.live/css/cssom/getComputedStyle-display-none-001.html) tho :)

**Comment by andruud** [comment]
> Blink implements it partially: the testcase above is good, but adding this breaks it [...]

Thanks, landed a fix for this now. (https://crbug.com/1399755).

> [general issue]

The current spec intentionally has no special case directly on display:none, and instead delays dealing with that until container-query-evaluation time (via having a principle box or not). This is to get a "stable" container situation in display:none. I.e. the container (which is an element, not a layout box) is always the "intended" container, and not a "random" container outside of display:none.

A related point is that container selection intentionally doesn't care whether an element has a pricinple box or not. I.e. we can select the right container without building the layout tree. This is important for Blink, since we need to know if something is a container or not sufficiently early in the pipeline. (This may be besides the point of this issue, but who knows where the discussion will lead).

> Firefox skips #inner-container and thinks that the query container is #outer-container.

I think we should avoid this behavior for the reasons I said. (#outer-container is not the intended container).

> Maybe elements inside display: none subtrees should be considered to not have any container?

That on the other hand seems probably OK? Although gCS would be worse than it needs to be when only the querying element (and not the container) is in display:none. And it's possibly unfortunate for style queries. But at least it's stable.

**Comment by lilles** [comment]
> > Maybe elements inside display: none subtrees should be considered to not have any container?
> 
> That on the other hand seems probably OK? Although gCS would be worse than it needs to be when only the querying element (and not the container) is in display:none. And it's possibly unfortunate for style queries. But at least it's stable.

Yeah, I think it sounds counter-intuitive for style queries to stop them from having effect inside display:none when other styles match as usual.
